### PR TITLE
test: authentication package unit tests

### PR DESCRIPTION
**Pull request by OmarAlJarrah**
# Situation

At the moment of creating this PR, we do not have unit-tests that cover all components in our new `sdk-core`. The goal of this PR is to add unit-tests that test all files in the `com.expediagroup.sdk.core.authentication` package.
# Task

Add unit tests using Junit-5 and Mockk.
# Action
Added new unit-tests for the following classes:
* `BearerAuthenticationInterceptor`
* `BearerAuthenticationManager`
* `BearerTokenStorageTest`
* `TokenResponse`
* `Credentials`

# Testing
**NaN**

# Results
All files in the `com.expediagroup.sdk.core.authentication` package are now tested.

# Notes
**NaN**

### CODE REVIEW

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationInterceptorTest.kt [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+        verify(exactly = 1) { authManager.isTokenAboutToExpire() }
+        verify(exactly = 1) { authManager.getAuthorizationHeaderValue() }
+        verify(exactly = 0) { authManager.authenticate() }
+        verify(exactly = 1) { requestBuilder.addHeader("Authorization", customToken) }
+        verify(exactly = 1) { chain.proceed(authorizedRequest) }
+        assertNotNull(response)
+    }
+
+    @Test
+    fun `ensureValidAuthentication should not call authenticate if token is not about to expire`() {
```
Comment (Mohammad-Dwairi) [inline code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationInterceptorTest.kt:263]: let's avoid mentioning implementation details in the test name. Same for the next test function
```suggestion
    fun `should not call authenticate if token is not about to expire`() {
```

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt [hunk 2]
```diff
@@ -0,0 +1,552 @@
...
+
+    @AfterEach
+    fun tearDown() {
+        clearMocks(requestExecutor)
+        authenticationManager.clearAuthentication()
+    }
+
+
+    /**
+     * Test successful authentication flow.
```
Comment (Mohammad-Dwairi) [inline code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:52]: can we remove these docs strings, the test function name should be sufficient

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt [hunk 3]
```diff
@@ -0,0 +1,552 @@
...
+     */
+    @Test
+    fun `authenticate should store token on successful response`() {
```
Comment (Mohammad-Dwairi) [inline code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:55]:
```suggestion
    fun `should authenticate and store access token on successful response`() {
```
It's better to describe the expected behavior in abstraction without tying the test to a specific method name.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+        // Assert
+        assertEquals("Bearer first_token", authenticationManager.getAuthorizationHeaderValue())
+        verify(exactly = 1) { requestExecutor.execute(any()) }
+    }
+
+    /**
+     * Test authentication failure due to server error.
+     */
+    @Test
+    fun `authenticate should throw ExpediaGroupAuthException on failure response`() {
```
Comment (Mohammad-Dwairi) [inline code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:89]:
```suggestion
    fun `should throw ExpediaGroupAuthException on failure response`() {
```

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
+        }
+        assertTrue(exception.message!!.contains("Failed to parse"))
+        verify(exactly = 1) { requestExecutor.execute(any()) }
+    }
+
+    /**
+     * Test isTokenAboutToExpire when token is not about to expire.
+     */
+    @Test
+    fun `isTokenAboutToExpire should return false when token is valid`() {
```
Comment (Mohammad-Dwairi) [inline code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:161]:
```suggestion
    fun `should treat the stored token as a valid token when not expired`() {
```

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt [hunk 6]
```diff
@@ -0,0 +1,552 @@
...
+
+        // Assert
+        assertFalse(isAboutToExpire)
+    }
+
+    /**
+     * Test isTokenAboutToExpire when token is about to expire.
+     */
+    @Test
+    fun `isTokenAboutToExpire should return true when token is about to expire`() {
```
Comment (Mohammad-Dwairi) [inline code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:195]:
```suggestion
    fun `should treat the stored token as a invalid token if expired`() {
```

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+        // Assert
+        assertTrue(isAboutToExpire)
+        verify(exactly = 1) { spyManager.isTokenAboutToExpire() }
+    }
+
+    /**
+     * Test clearing authentication.
+     */
+    @Test
+    fun `clearAuthentication should reset the token`() {
```
Comment (Mohammad-Dwairi) [inline code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:238]:
```suggestion
    fun `should handle token clearance`() {
```

Can we follow the same pattern for the reset of test cases here?

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/TokenResponseTest.kt [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+import com.expediagroup.sdk.core.http.Status
+import com.expediagroup.sdk.core.model.exception.client.ExpediaGroupResponseParsingException
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+
+
+class TokenResponseTest {
+    @Test
+    fun `parsed instances accessToken value should map to api response`() {
```
Comment (Mohammad-Dwairi) [inline code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/TokenResponseTest.kt:18]:
```suggestion
    fun `should map to the expected api response`() {
```
We can then combine this test with the one below. Just add the assertion for the `expiresIn`

**Review by Mohammad-Dwairi — DISMISSED** [review]

File: code/build.gradle [hunk 9]
```diff
@@ -37,7 +37,8 @@ dependencies {
     /* Testing */
     testImplementation platform('org.junit:junit-bom:5.11.3')
     testImplementation 'org.junit.jupiter:junit-jupiter-api'
-    testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine'
+    testImplementation 'org.junit.jupiter:junit-jupiter-engine'
```
[the remark this answers is not in the document]
Comment (OmarAlJarrah) [reply code/build.gradle:40]: Updated.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationInterceptorTest.kt [hunk 10]
```diff
@@ -0,0 +1,361 @@
+package com.expediagroup.sdk.core.authentication.bearer
+
+import com.expediagroup.sdk.core.http.Request
+import com.expediagroup.sdk.core.interceptor.Interceptor
+import com.expediagroup.sdk.core.model.exception.service.ExpediaGroupAuthException
+import io.mockk.*
```
[the remark this answers is not in the document]
Comment (OmarAlJarrah) [reply code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationInterceptorTest.kt:6]: Done.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationInterceptorTest.kt (hunk 1 above)
Comment (OmarAlJarrah) [reply code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationInterceptorTest.kt:263]: Done.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt (hunk 2 above)
Comment (OmarAlJarrah) [reply code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:52]: done.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt (hunk 3 above)
Comment (OmarAlJarrah) [reply code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:55]: done.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt (hunk 4 above)
Comment (OmarAlJarrah) [reply code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:89]: done.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt (hunk 5 above)
Comment (OmarAlJarrah) [reply code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:161]: done.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt (hunk 6 above)
Comment (OmarAlJarrah) [reply code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:195]: done.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt (hunk 7 above)
Comment (OmarAlJarrah) [reply code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/BearerAuthenticationManagerTest.kt:238]: done.

File: code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/TokenResponseTest.kt (hunk 8 above)
Comment (OmarAlJarrah) [reply code/src/test/kotlin/com/expediagroup/sdk/core/authentication/bearer/TokenResponseTest.kt:18]: done.

**Review by Mohammad-Dwairi — APPROVED** [review]

**Review by mohnoor94 — APPROVED** [review]

**Pull request merged**